pkg/controller/jobs/rayjob/rayjob_controller.go

```diff
@@ -0,0 +1,284 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rayjob
+
+import (
+	"context"
+	"strings"
+
+	rayjobapi "github.com/ray-project/kuberay/ray-operator/apis/ray/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+)
+
+var (
+	gvk           = rayjobapi.GroupVersion.WithKind("RayJob")
+	FrameworkName = "ray.io/rayjob"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:  SetupIndexes,
+		NewReconciler: NewReconciler,
+		SetupWebhook:  SetupRayJobWebhook,
+		JobType:       &rayjobapi.RayJob{},
+		AddToScheme:   rayjobapi.AddToScheme,
+	}))
+}
+
+// RayJobReconciler reconciles a RayJob object
+type RayJobReconciler jobframework.JobReconciler
+
+func NewReconciler(
+	scheme *runtime.Scheme,
+	client client.Client,
+	record record.EventRecorder,
+	opts ...jobframework.Option) jobframework.JobReconcilerInterface {
+	return (*RayJobReconciler)(jobframework.NewReconciler(scheme,
+		client,
+		record,
+		opts...,
+	))
+}
+
+type RayJob rayjobapi.RayJob
+
+var _ jobframework.GenericJob = (*RayJob)(nil)
+
+func (j *RayJob) Object() client.Object {
+	return (*rayjobapi.RayJob)(j)
+}
+
+func (j *RayJob) IsSuspended() bool {
+	return j.Spec.Suspend
+}
+
+func (j *RayJob) IsActive() bool {
+	return j.Status.JobDeploymentStatus != rayjobapi.JobDeploymentStatusSuspended
+}
+
+func (j *RayJob) Suspend() {
+	j.Spec.Suspend = true
+}
+
+func (j *RayJob) ResetStatus() bool {
+	if j.Status.StartTime == nil {
+		return false
+	}
+	j.Status.StartTime = nil
+	return true
+}
+
+func (j *RayJob) GetGVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (j *RayJob) PodSets() []kueue.PodSet {
+	// len = workerGroups + head
+	podSets := make([]kueue.PodSet, len(j.Spec.RayClusterSpec.WorkerGroupSpecs)+1)
+
+	// head
+	podSets[0] = kueue.PodSet{
+		Name:     "head",
+		Template: *j.Spec.RayClusterSpec.HeadGroupSpec.Template.DeepCopy(),
+		Count:    1,
+	}
+
+	// workers
+	for index := range j.Spec.RayClusterSpec.WorkerGroupSpecs {
+		wgs := &j.Spec.RayClusterSpec.WorkerGroupSpecs[index]
+		replicas := int32(1)
+		if wgs.Replicas != nil {
+			replicas = *wgs.Replicas
+		}
+		podSets[index+1] = kueue.PodSet{
+			Name:     strings.ToLower(wgs.GroupName),
```

Comment:
Why do we need to make `groupName`  lower?

Reply:
Maybe, because use lowers in mpijob-controller?

https://github.com/kubernetes-sigs/kueue/blob/74f45f69612a2d36ace91ae455f1e9fdbcd53037/pkg/controller/jobs/mpijob/mpijob_controller.go#L116